ast/src/parsed/mod.rs

```diff
@@ -61,18 +66,192 @@ pub enum Expression<T> {
     ),
 }
 
+impl<T> From<ShiftedPolynomialReference<T>> for Expression<T> {
+    fn from(value: ShiftedPolynomialReference<T>) -> Self {
+        Self::PolynomialReference(value)
+    }
+}
+
 #[derive(Debug, PartialEq, Eq, Default, Clone)]
 pub struct PolynomialName<T> {
     pub name: String,
     pub array_size: Option<Expression<T>>,
 }
 
+/// A polynomial with an optional shift
 #[derive(Debug, PartialEq, Eq, Default, Clone)]
-pub struct PolynomialReference<T> {
-    pub namespace: Option<String>,
-    pub name: String,
-    pub index: Option<Box<Expression<T>>>,
-    pub next: bool,
+pub struct ShiftedPolynomialReference<T> {
+    /// Whether we shift or not
+    is_next: bool,
+    /// The underlying polynomial
+    pol: NamespacedPolynomialReference<T>,
+}
+
+impl<T> ShiftedPolynomialReference<T> {
+    /// Returns the underlying namespaced polynomial
+    pub fn into_namespaced(self) -> NamespacedPolynomialReference<T> {
+        self.pol
+    }
+
+    /// Returns the shift of this polynomial
+    pub fn shift(&self) -> bool {
+        self.is_next
+    }
```

Comment:
The comment doesn't quite fit the return type.

Reply:
So "shift" is meant in a general sense here, like `a'''''''` but we currently only support 0 and 1. I thought the interface could be general and the implementation specific

Reply:
the idea here is to be generic over any amounts of shifts: in PIL2 you'll also be able to shift by any constant. So while currently the shift is boolean (is_next), this function will still return "the shifted amount"